### PR TITLE
Fix type update behaviour in expression clones

### PR DESCRIPTION
**Pull request by mlange05**
_Note the problem this PR intends to fix is the underlying problem to the symptoms reported in #128. Many thanks to @rolfhm for providing the MFE that helped triage this._

This PR adds a test and fixes certain corner cases for type updates of symbols via `symbol.clone(type-<...>)`. The key problem is that there were edge-cases where we would preserve existing type info stored in the associated scope of the symbol, if the new type was deemed inadequate. This PR now simplifies this logic and only ignores the incoming type if it is `None`, as we do not allow `None` type info in the symbol table (`BaseType.DEFERRED` signifies unknown types). 

In addition, a test was added that pins down the intended behaviour for plain scalar symbols. In the future this could be extended to procedures etc., but this felt overkill for the current problem. 

I've also needed to change the scope handling logic in the fparser-based standalone expression parsing utility, where we can never gain type info (no declarations in expressions), but would wipe existing type info by setting it to DEFERRED. We now instead use a dummy to parse expressions and suppress that info in the subsequent rescoping (see inline comments).

### CODE REVIEW

File: loki/frontend/fparser.py [hunk 1]
```diff
@@ -138,9 +139,14 @@ def parse_fparser_expression(source, scope):
...
+
+    # We parse the standalone expression with a dummy scope, to
+    # avoid overriding existing type info from the given scope.
+    _ir = parse_fparser_ast(ast, source, scope=Scope())
+
     # TODO: use rescope visitor for this
-    rescope_map = {v: v.clone(scope=scope) for v in FindTypedSymbols().visit(_ir)}
+    # In the rescoping we force using the existing type info via `type=None`
+    rescope_map = {v: v.clone(scope=scope, type=None) for v in FindTypedSymbols().visit(_ir)}
     _ir = SubstituteExpressions(rescope_map).visit(_ir)
```
Comment (reuterbal) [inline loki/frontend/fparser.py:150]: Might make sense to do this by using corresponding functionality that's already available (and thus take care of the TODO in the process), e.g.:
```
{v: v.rescope(scope=scope)}
```
or, even better because the parse expression may actually be using symbols from a parent scope:
```
_ir = AttachScopes().visit(_ir, scope=scope)
```

File: tests/test_expression.py [hunk 2]
```diff
@@ -1157,6 +1157,38 @@ def test_variable_clone(initype, inireftype, newtype, newreftype):
...
+     SymbolAttributes(BasicType.DEFERRED)),
+])
+def test_variable_clone_type(initype, newtype, reftype):
+    """
+    Test type updates are handled as expected and types are never ``None``.
+    """
+    scope = Scope()
+    var = symbols.Variable(name='var', scope=scope, type=initype)
+    assert 'var' in scope.symbol_attrs
+    new = var.clone(type=newtype)
```
Comment (reuterbal) [inline tests/test_expression.py:1189]: This just needs a `# pylint: disable=no-member` annotation.
(We really ought to convert the `Variable` factory class to a plain function)

**Review by reuterbal — CHANGES_REQUESTED** [review]
This is very nice, many thanks! The fact that the logic in the symbol constructor could be that simple makes the current implementation almost embarrassing. 

There are still a few failing tests and I've left a comment on the expression parsing change.

File: loki/frontend/fparser.py (hunk 1 above)
Comment (mlange05) [reply loki/frontend/fparser.py]: Agreed, the second suggestion looks much neater and works very well. Updated.

**Comment by mlange05** [comment]
@reuterbal As discussed offline, this branch now includes a small, but subtle behaviour change to get the expression tests back up and running. For this we now treat a `Variable.clone(scope=scope)` call, where the variable previously had no scope attached, as a re-scope, and prefer the type info in the provided scope over the previously stored DEFERRED type. This requires a small change in the `TypeDef.variables` property, but once fixed I think this is actually good. 

So, please have another look and see if you agree. All other suggestions have been incorporated, let's see if CI can be convinced too.

**Comment by mlange05** [comment]
@reuterbal As discussed offline, this branch now includes a small, but subtle behaviour change to get the expression tests back up and running. For this we now treat a `Variable.clone(scope=scope)` call, where the variable previously had no scope attached, as a re-scope, and prefer the type info in the provided scope over the previously stored DEFERRED type. This requires a small change in the `TypeDef.variables` property, but once fixed I think this is actually good. 

So, please have another look and see if you agree. All other suggestions have been incorporated, let's see if CI can be convinced too.

File: loki/expression/symbols.py [hunk 3]
```diff
@@ -815,16 +819,10 @@ def __new__(cls, **kwargs):
         scope = kwargs.get('scope')
         _type = kwargs.get('type')
 
-        if scope is not None and (_type is None or _type.dtype is BasicType.DEFERRED):
+        if scope is not None and _type is None:
             # Try to determine stored type information if we have no or only deferred type
```
Comment (reuterbal) [inline loki/expression/symbols.py:823]: The comment is out of date. Given the complexity, it might be worth expanding this here a little bit in general?

**Review by reuterbal — APPROVED** [review]
Looks great now, many thanks for figuring this out! The comment in the `clone` method is a little out of date now (and factually wrong), so might be worth updating this.

File: loki/expression/symbols.py (hunk 3 above)
Comment (mlange05) [reply loki/expression/symbols.py]: Good spot. Should be fixed now. Thanks!
